keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,389 @@
+
+Users have requested that we should also allow them to configure the failure conditions under which the JobSet will be restarted,
+or when the JobSet should be failed immediately without going through the allowed number of restarts. 
+
+This will allow users to make more efficient use of their computing resources, by avoiding `maxRestarts` number of unnecessary 
+JobSet restarts in the case of a non-retriable failure (e.g., an application code bug), while still restarting in the event of
+retriable failures (e.g., a host maintenace event).
+
+This is especially important for large scale distributed training workloads on expensive hardware accelerators. Recreating the 
+entire JobSet may take ~1 minute or so (per latest scale tests on several node pools of ~3k nodes each), and if `maxRestarts`
+is set to a very high number to allow for disruptions (which occur often at a large scale like this), the workload could
+waste roughly `maxRestarts` number of minutes repeatedly recreating despite the fact that it is doomed to fail due to a non-retriable
+error like an application code bug. This is very expensive and wasteful on expensive, large scale clusters.
+
+We need to define a more configurable Failure Policy API which will allow users to define their own restart and failure semantics, 
+based on the type of child job failure occurring.
+
+### Goals
+
+- Enhance the Failure Policy API with options allowing the user to control the restart and failure semantics of the JobSet.
+
+### Non-Goals
+
+- Handling every possible type of restart and failure semantics. We just want to provide enough flexibility to handle common use
+cases.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, in order to use my computing resources as efficiently as possible, I want to 
+configure my JobSet to restart in the event of a child job failure due to a retriable error
+like host maintance, but to fail the JobSet immediately without any unnecessary restart 
+attempts in the event of an non-retriable application code error. 
+
+**Example Failure Policy Configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: configurable-failure-policy-1
+spec:
+  # Failure Policy configured to fail the JobSet immediately if a job failure reason for any child job was NOT due to a SIGTERM (e.g. host maintenance).
+  # Otherwise, restart up to 10 times.
+  failurePolicy:
+    rules:
+    - action: FailJobSet
+      onJobFailureReasons: 
+      - PodFailurePolicy
+    maxRestarts: 10
+  replicatedJobs:
+  - name: buggy-job 
+    replicas: 1
+    template:
+      spec:
+        parallelism: 1
+        completions: 1
+        backoffLimit: 0
+        # podFailurePolicy which fails job immediately if job was not killed by SIGTERM (i.e., graceful node shutdown for maintenance events)
+        podFailurePolicy:
+          rules:
+          - action: FailJob
```

Comment:
Did you find that you don't need `onPodConditions` here after all?

Reply:
I actually do need to set it to `onPonConditions: []` to pass validation still, but I omitted it because I thought that unsightly bug might confuse readers and distract from the KEP 😆

Reply:
KEPs are free documentation for the features so I like to make the user stories as real test cases. So users could copy these and try out the feature. But that is just me.

Reply:
Why is an empty list required? Does the absence of the field not indicate that?

Reply:
It's a bug in one of the k8s api packages we are using, haven't tracked it down yet